### ISSUE TITLE
Mempool admission inconsistent with chaining checks

### ISSUE THREAD

**Issue by kantai** [blockstack/stacks-blockchain#2150]
The `BadNonce` error response to the transaction push API on the `stacks-node` supplies the caller with the current max nonce in the mempool. However, in the presence of forks, this is a very misleading response (see discussion in https://github.com/blockstack/stacks-blockchain-api/issues/359).

If we don't want the stacks-node mempool to automatically try to apply transactions across forks, then a client needs a way to know that the transactions previously included in the mempool are now not being processed because they are in a different fork. A client that receives a BadNonce error, which is the problem that the client is seeing (not ConflictingNonceInMempool), needs to be able to figure out what the correct nonce should be. The mempool is capable of answering this question (and it even tries to in the error response), but it currently answers it wrong, in core::mempool:

```rust
    fn get_next_nonce_as_participant_type(
        conn: &DBConn,
        addr: &StacksAddress,
        as_origin: bool,
    ) -> Result<u64, db_error> {
        let sql = format!(
            "SELECT ifnull(max({0}_nonce + 1), 0) FROM mempool WHERE {0}_address = ?1",
            if as_origin { "origin" } else { "sponsor" }
        );
        match conn.query_row_and_then(&sql, &[addr.to_string()], |row| u64::from_row(row)) {
            Ok(max) => Ok(max),
            Err(db_error::SqliteError(SqliteError::QueryReturnedNoRows)) => Ok(0),
            otherwise => otherwise,
        }
    }

    pub fn get_next_nonce_for_address(
        conn: &DBConn,
        address: &StacksAddress,
    ) -> Result<u64, db_error> {
        let as_origin = MemPoolDB::get_next_nonce_as_participant_type(conn, address, true)?;
        let as_sponsor = MemPoolDB::get_next_nonce_as_participant_type(conn, address, false)?;
        Ok(cmp::max(as_origin, as_sponsor))
    }
```

That returns the max nonce across all forks, which isn't useful to a client during such a fork. I think this really should return the max nonce _in the canonical fork_, because that's ultimately how the tx is going to be validated.

**Comment by kantai** [comment]
On looking more into this, it already is the case that we're returning an appropriate response for `BadNonce`. Looking at the logs:

```
    Reject transaction be808faa1a62dfbe117f75b99d32a04edd6ae1c9ed8a7e4ea4e15419522cf6a6: BadNonces(TransactionNonceMismatch { expected: 112, actual: 111, txid: be808faa1a62dfbe117f75b99d32a04edd6ae1c9ed8a7e4ea4e15419522cf6a6, principal: Standard(StandardPrincipalData(26, [22, 66, 71, 214, 242, 180, 37, 172, 87, 113, 66, 58, 230, 200, 12, 117, 79, 113, 114, 176])), is_origin: true, quiet: true })
```

The BadNonces response _is_ providing a suggestion, and that suggestion is sort of correct. I think what's happening here is a little more subtle.

The bug here is that the mempool admission does the following:

1. See if the transaction can be admitted to the current chain state.
2. If not, then check the mempool for the highest nonces for the addresses in the tx (this step is buggy in the event of forks -- it considers the highest nonce across all forks).

Instead, this should:

1. See if the transaction can be admitted to the current chain state.
2. If not:
    a. Check if the nonces are within the allowed chaining limit, and if so, allow them to be passed to the mempool `try_add_tx`.
    b. If nonces are > chaining limit, throw a TooMuchChaining error
    c. If nonces are < chaining limit, throw the BadNonces error, which includes the correct suggested nonce.
3. Let the mempool `try_add_tx` handle the "replace-by-fee" or "replace-across-fork" checks (as it currently does).

**Comment by kantai** [comment]
Closed via #2153
